refactor: improvements to log handling

datawhores opened the pull request:
1. while richhandler work well for logging to the console, it formatting for file logs is not clean. 

![image](https://github.com/user-attachments/assets/139917f0-2687-4b65-af51-5896ff3aba63)
![image](https://github.com/user-attachments/assets/65591a40-5d8a-46dd-a3f1-2b02fb193b00)

As can be seen in the photo above, many of the lines are not uniform, which makes it harder to read the log file

Python already has a filehandler, which is able to output logs that are much more readable
![image](https://github.com/user-attachments/assets/c74caac8-5c2d-4722-8039-c89a212823ba)

 
2. Allow logs to be in seperate folders for each type

NTFSvolume commented:
The issue is with the text editor you are using. It's word wrapping on every word that has a tab after it, putting it on the next line.

I can't identify which editor you are using from the image

![image](https://github.com/user-attachments/assets/b5b2b32f-4896-42c9-98ea-325eb63f5b2c)

datawhores commented:
> The issue is with the text editor you are using. It's word wrapping on every word that has a tab after it, putting it on the next line.
> 
> I can't identify which editor you are using from the image
> 
> ![image](https://private-user-images.githubusercontent.com/172021377/391300088-b5b2b32f-4896-42c9-98ea-325eb63f5b2c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMwMTc3MzgsIm5iZiI6MTczMzAxNzQzOCwicGF0aCI6Ii8xNzIwMjEzNzcvMzkxMzAwMDg4LWI1YjJiMzJmLTQ4OTYtNDJjOS05OGVhLTMyNWViNjNmNWIyYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIwMVQwMTQzNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMGQyYTUzODc0OWM1MWNkZDdlMThhNmY0YjE3ZTQzYmQ0Njk1NDJhMmIzYjNlYWY5MDIwYzlkMzQwMzk0MjVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.hzBzXaVDiChvX5UAgimap7Py1TLxQbcjlBnFKNP817E)

It just on linux, but the same issue has applied to xed and visual studio code

datawhores commented:
I added some more visualizations of spacing and new lines

![image](https://github.com/user-attachments/assets/fa47bd6f-e859-424c-90bc-246d17a7f7a9)
"." are for a single space
"down arrow" is for new line character

NTFSvolume commented:
Don't really know what's causing it. It looks fine for me on VSCode, popOS 22.04. But i'm pretty sure its something obvious that we are just missing.

![Screenshot from 2024-11-30 20-59-00](https://github.com/user-attachments/assets/187211c2-9897-424e-95ca-197f8b8031a1)

NTFSvolume commented:
I think this may be it. A possible issue with the terminal emulator your are using: https://github.com/Textualize/rich/issues/1019#issuecomment-781957963

datawhores commented:
I saw this issue
https://github.com/Textualize/rich/discussions/344

I changed the width which helped, but I see extra padding like in the issue

NTFSvolume commented:
Both console objects for the logs use a file and have a fixed width

https://github.com/jbsparrow/CyberDropDownloader/blob/83ee9a0f7ce9216b7fa9df3e9afc99c6a01677d5/cyberdrop_dl/main.py#L167-L175

The default width is 240 which should be enough

https://github.com/jbsparrow/CyberDropDownloader/blob/83ee9a0f7ce9216b7fa9df3e9afc99c6a01677d5/cyberdrop_dl/utils/constants.py#L10

I will make a VM to see if i can replicate the issue

NTFSvolume commented:
I just noticed that the issue is also happening when you used the regular Python FileHandler. Notice that `logger` in the first 2 lines is either cut off or wrapped to the next line. It's like the file has a fixed width for some reason... 

Will try to reproduce on my end.

NTFSvolume commented:
> I added some more visualizations of spacing and new lines
> 
> ![image](https://private-user-images.githubusercontent.com/67020411/391301102-fa47bd6f-e859-424c-90bc-246d17a7f7a9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMwMjU5NDQsIm5iZiI6MTczMzAyNTY0NCwicGF0aCI6Ii82NzAyMDQxMS8zOTEzMDExMDItZmE0N2JkNmYtZTg1OS00MjRjLTkwYmMtMjQ2ZDE3YTdmN2E5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjAxVDA0MDA0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJmZWM4MzU4MjhkMGQ5MGFkODQzZDQzMDcwNjQ4NDZjOWY4MjUwZjQwMGY3NDljMDBmNWUzYmI0NTU2NzIzMjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gty5SWhjA9Lk3bfClNz-4ChI58X3B0zlgwdqasLMCcI) "." are for a single space "down arrow" is for new line character

Wait, never mind, that IS vscode line wrapping. Look at the line numbers. You just have to disable line wrapping in vscode

NTFSvolume commented:
Select `view` -> `word wrap` or press `Alt + Z`

datawhores commented:
@NTFSvolume 
Removing the word wrapping seems  to fix the issue

I notice that the program has the width of the screen set to 240, I'm not sure why this number was picked. Maybe based on Jules max screen size. However, on my screen when I maximize my screen it is only 238 characters. I actually played around with setting the value to 238, and with that the word wrapping issue was not present. 

![image](https://github.com/user-attachments/assets/04f4c760-abbc-4315-a86a-5922d8e0e9f8)

NTFSvolume commented:
> @NTFSvolume Removing the word wrapping seems to fix the issue
> 
> I notice that the program has the width of the screen set to 240, I'm not sure why this number was picked. Maybe based on Jules max screen size. However, on my screen when I maximize my screen it is only 238 characters. 

I was the one who picked 240. It's 240 characters, not pixels. I initially picked 160 but some long URLs did not fit on a single line, that's why i increased it. 

240 was long enough to fit long URLs like bunkr's with their UUID but still show the full line without having to scroll left or right on a 1366x768 monitor, which is a nice lower bound of modern display resolutions. It only cutoffs the end, the `logger.py:<line_number>`, but that not really relevant.

> I actually played around with setting the value to 238, and with that the word wrapping issue was not present. 
> Basically the logs are much cleaner and work with word wrap on or off if I reduce the width value. 

Changing it to 238 it fine. It's not a big difference.

> Getting the maximum width number is not really easy from a user perspective though
> 
> I have to maximize a terminal and run
> 
> ```
> echo $COLUMNS
> ```
> 
> or
> 
> ```
> tput cols
> ```

We can not get the width from the terminal. If the terminal is resized at any point, the formatting of the logs file gets messed up. The width cannot be too big neither. If an user has an ultrawide or superultrawide monitor, the logs will be a pain to read on a normal screen.

It needs to be a defined constant width for every user. Supporting word wrapping is not relevant. It just a feature on text editors and some do not even have it.

datawhores commented:
I changed it to 180 system

Everything fits without having to scroll to the left
I added an option to change this on the user system with a min of 50

NTFSvolume commented on cyberdrop_dl/config_definitions/config_settings.py:
```diff
@@ -44,6 +44,8 @@ class Logs(AliasModel):
     download_error_urls_filename: NonEmptyStr = "Download_Error_URLs.csv"
     scrape_error_urls_filename: NonEmptyStr = "Scrape_Error_URLs.csv"
     rotate_logs: bool = False
+    seperate_folders: bool = False
```
Instead of adding a new option, change the existing log file name options to take a `path`. If the user wants to use a subfolder, they can use something like `Download_Errors_Logs/Download_Error.csv`. This will also allow them to save all the logs of all the configs on the same folder by using an absolute path instead of a relative path, but they will have to also set  `rotate_logs` to make sure the do not override each other

NTFSvolume commented on cyberdrop_dl/config_definitions/config_settings.py:
```diff
@@ -44,6 +44,8 @@ class Logs(AliasModel):
     scrape_error_urls_filename: NonEmptyStr = "Scrape_Error_URLs.csv"
     rotate_logs: bool = False
+    seperate_folders: bool = False
+    file_log_line_width_limit: PositiveInt = Field(default=240, ge=50, required=True)
```
`default` and `required` are mutually exclusive. Leave just `default`

NTFSvolume replied:
Also try to use a shorter name like `log_line_width`

NTFSvolume commented on cyberdrop_dl/utils/logger.py:
```diff
@@ -60,3 +60,9 @@ def _log_to_console(level: int, record: str, *_, **__) -> None:
         console.log(record)
+
+
+def get_log_formatter():
+    return logging.Formatter(
+        "[%(asctime)s]      %(levelname)-8s%(message)s     %(filename)s:%(lineno)d\n", datefmt="%m/%d/%y %H:%M:%S"
+    )
```
This function is not used

datawhores commented:
Tested with both the
original log settings as a filename
new log setting as relative filepath
new log setting as full path, the log folder settings is just ignore
